lgc/patch/PatchWorkarounds.cpp

```diff
@@ -0,0 +1,141 @@
+ ***********************************************************************************************************************
+ */
+
+#include "PatchWorkarounds.h"
+#include "lgc/BuilderBase.h"
+#include "lgc/state/PipelineShaders.h"
+#include "lgc/state/PipelineState.h"
+#include "lgc/state/TargetInfo.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lgc-patch-workarounds"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Define static members (no initializer needed as LLVM only cares about the address of ID, never its value).
+char PatchWorkarounds::ID;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of LLVM patching operations for peephole optimizations.
+FunctionPass *createPatchWorkarounds() {
+  return new PatchWorkarounds();
+}
+
+PatchWorkarounds::PatchWorkarounds() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will add workarounds.
+bool PatchWorkarounds::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Workarounds\n");
+
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(function.getParent());
+  m_context = &function.getContext();
+  m_builder = std::make_unique<IRBuilder<>>(*m_context);
+  m_processed.clear();
+
+  m_changed = false;
+
+  // At present we only apply a single workaround in this pass. Makes sense to check here an early exit the pass if not
+  // required
+  if (!m_pipelineState->getOptions().disableImageResourceCheck &&
+      m_pipelineState->getTargetInfo().getGpuWorkarounds().gfx10.waFixBadImageDescriptor)
+    visit(function);
+
+  return m_changed;
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// @param callInst : The instruction
+void PatchWorkarounds::visitCallInst(CallInst &callInst) {
+  Function *const calledFunc = callInst.getCalledFunction();
+  if (!calledFunc)
+    return;
+
+  auto name = calledFunc->getName();
+
+  const bool isImage = name.startswith("llvm.amdgcn.image.");
+
+  // A buffer descriptor may be incorrectly given when it should be an image descriptor, we need to fix it to valid
+  // buffer type (0) to make hardware happily ignore it. This is to check and fix against buggy applications which
+  // declares a image descriptor in shader but provide a buffer descriptor in driver. Note this only applies to gfx10.1.
+  if (isImage) {
```

Comment:
nit: use an early return?
I can imagine that there will be more workarounds added here in the future, but in that case it might be cleaner to for example add specialized `visitCallSomethingInst` functions and make this generic function simply dispatch to them.

Reply:
Yes, I probably should write it with more workarounds being added in future as an assumption.

Reply:
I've rewritten this now - see if this looks any better.
Policy now is to iterate over appropriate function uses (at the module level) and process any call instructions that match the pattern.